### PR TITLE
[BUG] fix (tab) indentation within CodeMirror

### LINKED ISSUE #4830: [BUG] Tab key doesn't seem to work in Hedy anymore?
Maybe this is just me and my computer, but the tab key on my keyboard doesn't seem to work anymore when typing code in Hedy. If i want to use indentation I need to type 4 spaces instead of one tab. When i use tab I select the variables label in the output screen. 

If it's just me, the issue can be closed (if you know a solution, I'd still like to know). If not, we might want to come up with a solution.

### PR DESCRIPTION
**Pull request by hasan-sh**
Fixes #4830 

**How to test?**
- go to code editor
- try indenting with tab (tip: shift-tab removes indentation as well!)
- and within blocks; e.g., now when you click `Enter` within a block (i.e., if or for), the correct indentation will be chosen based on the previous and next lines. 
```python
for i in range 1 to 5 # (CLICK ENTER)
    # CURSOR_STARTSHERE
    print i 

or 

for i in range 1 to 5
    print i #  (CLICK ENTER)
    # CURSOR_STARTS_HERE
```

### CODE REVIEW

File: static/js/cm-editor.ts [hunk 1]
```diff
@@ -117,7 +121,17 @@ export class HedyCodeMirrorEditor implements HedyEditor {
...
                 ]),
+                indentUnit.of(indentSize),
+                indentService.of((context, pos) => {
+                    const previousLine = context.lineAt(pos, -1);
+                    const nextLine = context.lineAt(pos + 1, -1);
+                    const nextIndentationSize =  nextLine.text.match(/^\s*/)![0].length;
+                    const prevIndentationSize =  previousLine.text.match(/^\s*/)![0].length;
+                    const indentBy = Math.max(prevIndentationSize, nextIndentationSize);
+                    return indentBy
+                }),
```
Comment (jpelay) [inline static/js/cm-editor.ts:134]: When there's no next line, this returns an error. Also, can you add this code to its own function?

**Review by jpelay — DISMISSED** [review]
Hello Hasan, what a nice and clever solution! I have a couple of comments:

With respect to indent with tab, we decided that it was better to be as accessible as possible and not trapping users in the editor on the case that they don't use pointing devices.

With respect to the approach, this is clever! But I think we can do a bit better. We actually  can automatically add indentation when the CodeMirror parser detects a certain type of command, i.e: if the user writes: `repeat 3 times` and hits enter, the next line will be indented, so, no need to even indent manually!

This would be a little trickier to implement, since it involves messing with the grammars and the tokenizer, but here is a simple example of how this can be done: https://lezer.codemirror.net/examples/indent/ and you can try this by going here: https://codemirror.net/try/ and changing the code to this:

```javascript
import {basicSetup, EditorView} from "codemirror"
import {python} from "@codemirror/lang-python"

new EditorView({
  doc: "print('hello')\n",
  extensions: [basicSetup, python()],
  parent: document.body
})
```
Then you can mess around with the indentation.

Do think you could do that approach? If not, we can always ship your solution and make this improvement later on!!! This is already a lot better over what we have :smile:

File: static/js/cm-editor.ts [hunk 2]
```diff
@@ -117,7 +121,17 @@ export class HedyCodeMirrorEditor implements HedyEditor {
                     ...defaultKeymap,
                     ...searchKeymap, // we need to replace this with our own search widget
                     ...historyKeymap,
+                    indentWithTab,
```
[the remark this answers is not in the document]
Comment (hasan-sh) [reply static/js/cm-editor.ts:124]: See this https://github.com/hedyorg/hedy/issues/4830#issuecomment-1838223040

File: static/js/cm-editor.ts (hunk 1 above)
Comment (hasan-sh) [reply static/js/cm-editor.ts]: Good one, fixed!

File: static/js/cm-editor.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpelay) [reply static/js/cm-editor.ts:125]: If we do this, then we need to mention somewhere that this escape hatch is possible! Perhaps on the tutorial?

[the remark this answers is not in the document]
Comment (hasan-sh) [reply static/js/cm-editor.ts:125]: I think users will know that naturally, but I do agree, that would be the perfect spot to include it.

**Comment by Felienne** [comment]
> * and within blocks; e.g., now when you click `Enter` within a block (i.e., if or for), the correct indentation will be chosen based on the previous and next lines.

Well, you say correct, but you mean the indentation of the line before, I think. Let's discuss this in the next meeting, I think there might be some *learning* value in doing it explicitly (esp in earlier levels) but we can totally merge this as this is what we had in Ace too!

**Comment by hasan-sh** [comment]
> Also, let's indeed add this to the manual (for teachers) and docs (for developers). Can you pick that up @hasan-sh?

Sure, will do in this PR.



> > * and within blocks; e.g., now when you click `Enter` within a block (i.e., if or for), the correct indentation will be chosen based on the previous and next lines.
> 
> Well, you say correct, but you mean the indentation of the line before, I think. Let's discuss this in the next meeting, I think there might be some _learning_ value in doing it explicitly (esp in earlier levels) but we can totally merge this as this is what we had in Ace too!

Yeah the current line or the next line indeed. So, this is really not considering the commands, but only the indentation of the current and next line w.r.t. where the cursor is.

**Comment by Felienne** [comment]
> Sure, will do in this PR.

Thanks! Ping me if that is ready so I can merge!

> So, this is really not considering the commands, but only the indentation of the current and next line w.r.t. where the cursor is.

No problem, that is what we have now too. We can discuss in the meeting how/if we want to make it smarter

**Comment by jpelay** [comment]
> I am not sure, reading the code, if this PR already adds the "press escape to exit the editor" hatch?

Yes it does! Hasan added it by applying an extension from CodeMirror that adds this as well.

**Comment by Felienne** [comment]
> > I am not sure, reading the code, if this PR already adds the "press escape to exit the editor" hatch?
> 
> Yes it does! Hasan added it by applying an extension from CodeMirror that adds this as well.

In that case we can merge this right away!

**Review by Felienne — APPROVED** [review]

**Pull request merged**